### PR TITLE
Add preCICE xSDK support

### PR DESCRIPTION
**Pull request by fsimonis**
This PR
* adds support for the build system changes introduced by the adaptation of the xSDK installation policies,
* fixes some python related bugs discovered during the migration, and
* fixes unprecice cmake and eigen dependencies.

### CODE REVIEW

File: var/spack/repos/builtin/packages/precice/package.py [hunk 1]
```diff
@@ -33,38 +33,68 @@ class Precice(CMakePackage):
...
+    depends_on('cmake@3.10.2:', type='build', when='@1.4:')
     depends_on('boost@1.60.0:')
     depends_on('boost@1.65.1:', when='@1.4:')
     depends_on('eigen@3.2:')
+    depends_on('eigen@:3.3.7', type='build', when='@:1.5')  # bug in prettyprint
     depends_on('libxml2')
     depends_on('mpi', when='+mpi')
     depends_on('petsc@3.6:', when='+petsc')
-    depends_on('python@2.7', when='+python', type=('build', 'run'))
+    depends_on('python@2.7.16', when='+python', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/precice/package.py]: It doesn't work with any other version of Python? Python 2 will no longer be supported by the end of the year.

File: var/spack/repos/builtin/packages/precice/package.py [hunk 2]
```diff
@@ -33,38 +33,68 @@ class Precice(CMakePackage):
...
+    depends_on('py-numpy@1.16.4', when='+python', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/precice/package.py]: It doesn't work with any other version of numpy?

File: var/spack/repos/builtin/packages/precice/package.py [hunk 3]
```diff
@@ -33,38 +33,68 @@ class Precice(CMakePackage):
...
 
+        libxml2_includes = spec['libxml2'].headers.directories[0]
         cmake_args = [
-            '-DMPI:BOOL=%s' % variant_bool('+mpi'),
-            '-DPETSC:BOOL=%s' % variant_bool('+petsc'),
-            '-DPYTHON:BOOL=%s' % variant_bool('+python'),
             '-DBUILD_SHARED_LIBS:BOOL=%s' % variant_bool('+shared'),
+            '-DMPI:BOOL=%s' % variant_bool('+mpi'),
+
+            '-DTPL_ENABLE_BOOST=ON' if xsdk_mode else None,
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/precice/package.py]: Does this add a `None` to the list? Might be better to conditionally append to the list.

File: var/spack/repos/builtin/packages/precice/package.py (hunk 2 above)
Comment (fsimonis) [reply var/spack/repos/builtin/packages/precice/package.py]: If I don't fix the numpy version, I get the following error:

<details>
<summary> click to expand </summary>

```
==> Error: An unsatisfiable version constraint has been detected for spec:

    python@2.7.16%gcc@9.1.0+bz2+ctypes+dbm+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=linux-ubuntu16.04-x86_64
        ^bzip2
            ^diffutils
        ^expat
        ^gdbm
            ^readline
                ^ncurses
                    ^pkgconf@1.6.1%gcc@9.1.0 arch=linux-ubuntu16.04-x86_64
        ^gettext@0.19.8.1%gcc@9.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu16.04-x86_64
            ^libxml2@2.9.9%gcc@9.1.0~python arch=linux-ubuntu16.04-x86_64
                ^libiconv@1.15%gcc@9.1.0 arch=linux-ubuntu16.04-x86_64
                ^xz@5.2.4%gcc@9.1.0 arch=linux-ubuntu16.04-x86_64
                ^zlib@1.2.11%gcc@9.1.0+optimize+pic+shared arch=linux-ubuntu16.04-x86_64
            ^tar
        ^libffi
        ^openssl
            ^perl@5.14.0:
        ^sqlite@3.0.8:


while trying to concretize the partial spec:

    py-numpy@1.17.0%gcc@9.1.0+blas+lapack arch=linux-ubuntu16.04-x86_64

py-numpy requires python version 3.5:, but spec asked for 2.7.16
```
</details>

File: var/spack/repos/builtin/packages/precice/package.py (hunk 1 above)
Comment (fsimonis) [reply var/spack/repos/builtin/packages/precice/package.py]: Currently we only test python 2, however, we plan to upgrade to python 3 soon for obvious reasons.
Another reason for fixing the python version is to simplify locating the numpy headers here:

https://github.com/spack/spack/pull/12356/files#diff-8b6a38871241552be40b51937352aad7R89-R90

File: var/spack/repos/builtin/packages/precice/package.py (hunk 3 above)
Comment (fsimonis) [reply var/spack/repos/builtin/packages/precice/package.py]: I was considering that too, however, it makes this section less readable. I'll look into it again.

Comment (fsimonis) [reply var/spack/repos/builtin/packages/precice/package.py]: I agree that it is a bit strange. I pushed a clean-up commit.

File: var/spack/repos/builtin/packages/precice/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/precice/package.py]: If any version of Python 2 works, you want:
```suggestion
    depends_on('python@:2', when='+python', type=('build', 'run'))
```
If only Python 2.7 works, you want:
```suggestion
    depends_on('python@2.7:2.8', when='+python', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/precice/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/precice/package.py]: In that case you want:
```suggestion
    # numpy 1.17+ requires Python 3
    depends_on('py-numpy@:1.16', when='+python', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/precice/package.py [hunk 4]
```diff
@@ -33,38 +33,78 @@ class Precice(CMakePackage):
...
+            ])
+        else:
+            cmake_args.append('-DPETSC:BOOL=OFF')
+
+        # Python
+        if '+python' in spec:
+            python_library = spec['python'].libs[0]
+            python_include = spec['python'].headers.directories[0]
+            numpy_include  = spec['py-numpy'].prefix.lib.join(
+                'python2.7/site-packages/numpy/core/include')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/precice/package.py]: To make this work for any Python version, you can do:
```python
numpy_include = join_path(
    spec['py-numpy'].prefix, 
    spec['python'].package.site_packages_dir, 
    'numpy', 'core', 'include')
```

**Pull request merged**